test/hotspot/jtreg/gc/g1/TestSkipRebuildRemsetPhase.java

```diff
@@ -0,0 +1,91 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package gc.g1;
+
+/*
+ * @test TestSkipRebuildRemsetPhase
+ * @summary Skip Rebuild Remset Phase if the Remark pause does not identify any rebuild candidates.
+ *          Fill up a region to above the set G1MixedGCLiveThresholdPercent.
+ * @requires vm.gc.G1
+ * @library /test/lib
+ * @build sun.hotspot.WhiteBox
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller sun.hotspot.WhiteBox
+ * @run driver gc.g1.TestSkipRebuildRemsetPhase
+ */
+
+import java.util.regex.Pattern;
+import java.util.regex.Matcher;
+
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.Asserts;
+import sun.hotspot.WhiteBox;
+
+public class TestSkipRebuildRemsetPhase {
+    public static void main(String[] args) throws Exception {
+        ProcessBuilder pb = ProcessTools.createJavaProcessBuilder("-Xbootclasspath/a:.",
+                                                                  "-XX:+UseG1GC",
+                                                                  "-XX:+UnlockExperimentalVMOptions",
+                                                                  "-XX:+UnlockDiagnosticVMOptions",
+                                                                  "-XX:+WhiteBoxAPI",
+                                                                  "-XX:G1MixedGCLiveThresholdPercent=20",
+                                                                  "-Xlog:gc+marking=debug,gc+phases=debug,gc+remset+tracking=trace",
+                                                                  "-Xms10M",
+                                                                  "-Xmx10M",
+                                                                  GCTest.class.getName());
+        OutputAnalyzer output = new OutputAnalyzer(pb.start());
+        output.shouldContain("Skipping Remembered Set Rebuild.");
+        output.shouldContain("No Remembered Sets to update after rebuild");
+        output.shouldHaveExitValue(0);
+    }
+
+    public static class GCTest {
+        public static void main(String args[]) throws Exception {
+            WhiteBox wb = WhiteBox.getWhiteBox();
+            // Allocate some memory less than region size.
+            Object used = alloc();
+
+            // Trigger the full GC using the WhiteBox API.
+            wb.fullGC();  // full
+
+            // Memory objects have been promoted to old by full GC.
+            // Concurrent cycle should not select any regions for rebuilding
+            wb.g1StartConcMarkCycle(); // concurrent-start, remark and cleanup
+
+            // Sleep to make sure concurrent cycle is done
+            while (wb.g1InConcurrentMark()) {
+                Thread.sleep(1000);
+            }
+
+            System.out.println(used);
+        }
+
+        private static Object alloc() {
+            // G1MixedGCLiveThresholdPercent=20
```

Comment:
Maybe this comment could be elaborated a bit further - i.e. "Since G1Mixed...ent is 20%, make sure to allocate something that is larger than that so that it will not be collected and the expected message printed."
Or just remove it.

Reply:
I think it is better to elaborate than remove.